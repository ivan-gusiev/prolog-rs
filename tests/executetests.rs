extern crate insta;
extern crate parameterized;
extern crate prolog_rs;

#[cfg(test)]
mod executetests {
    use insta::assert_display_snapshot;
    use parameterized::parameterized;
    use prolog_rs::{
        compile::{compile_program, compile_query},
        data::CodePtr,
        lang::parse_struct,
        symbol::SymbolTable,
        util::{case, lbl_for, run_just_query, writeout_sym},
        Machine, var::{VarInfo, VarBindings},
    };

    #[parameterized(input = {
        "p(Z, h(Z,W), f(W))",
        "f(X, g(X,a))",
        "f(b, Y)",
        "a"
    })]
    fn test_query_execute(input: &str) {
        let mut symbol_table = SymbolTable::new();
        let query = parse_struct(input, &mut symbol_table).unwrap();

        let labels = lbl_for(query.functor());
        let query_result = compile_query(query, &labels).unwrap();
        let mut machine = Machine::new();
        run_just_query(&mut machine, &query_result.instructions).expect("machine failure");

        let output = writeout_sym(
            &machine.iter_heap().copied().collect::<Vec<_>>(),
            &symbol_table,
        );
        assert_display_snapshot!(case(input, output));
    }

    #[parameterized(input = {
        ("p(Z, h(Z,W), f(W))", "p(f(X), h(Y, f(a)), Y)"),
        ("f(b, Y)", "f(X, g(X,a))"),
        ("f(X, g(X,a))", "p(f(X), h(Y, f(a)), Y)"),
        ("h(l(p(A, Y), p(B, Y)))", "h(l(p(u, v), p(w, H)))"),
    })]
    fn test_program_execute(input: (&str, &str)) {
        let (query_text, program_text) = input;
        let mut symbol_table = SymbolTable::new();
        let query = parse_struct(query_text, &mut symbol_table).unwrap();
        let program = parse_struct(program_text, &mut symbol_table).unwrap();
        let mut machine = Machine::new();

        let labels = lbl_for(query.functor());
        let query_result = compile_query(query, &labels).unwrap();
        let mut query_bindings = VarBindings::default();

        let program_result = compile_program(program);
        machine.set_code(&program_result.instructions);
        let p = machine.append_code(&query_result.instructions);
        machine.set_p(p);
        machine.execute().with_call_hook(|m| {
            m.bind_variables(&query_result.var_mapping).map(|vars| query_bindings = vars)
        }) .run().expect("machine failure");

        let input = format!("({query_text}, {program_text})");
        let output = if machine.get_fail() {
            machine.dbg(&symbol_table)
        } else {
            let program_bindings = machine
                .bind_variables(&program_result.var_mapping)
                .expect("decompile failure");

            format!(
                "{}\n{}\n{}",
                machine.dbg(&symbol_table),
                writeout_sym(
                    &machine.describe_vars(&query_bindings).unwrap(),
                    &symbol_table
                ),
                writeout_sym(
                    &machine.describe_vars(&program_bindings).unwrap(),
                    &symbol_table
                )
            )
        };
        assert_display_snapshot!(case(input, output));
    }
}
