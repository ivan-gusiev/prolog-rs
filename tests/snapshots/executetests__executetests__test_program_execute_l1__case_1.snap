---
source: tests/executetests.rs
expression: "case(input, output)"
---
(f(b, Y), f(X, g(X,a)))
-----
h: HEAP[9]
s: HEAP[2]
p: CODE[10]
mode: write
fail: false
halt: false
code:
000	get_variable X3, X1
001	get_structure g/2, X2
002	unify_value X3
003	unify_variable X4
004	get_structure a/0, X4
005	proceed
006	allocate 1
007	put_structure b/0, X1
008	put_variable Y1, X2
009	call @0

heap:
000	<STR,1>
001	b/0
002	<STR,4>
003	<STR,4>
004	g/2
005	<STR,1>
006	<STR,8>
007	<STR,8>
008	a/0

regs:
000	<EMPTY>
001	<STR,1>
002	<REF,2>
003	<STR,1>
004	<REF,6>

pdl:

stack depth 0, cp=CODE[0]:
000	<REF,2>


000	Y	(HEAP[2]) =	<STR,4>	// g(b, a)

000	X	(HEAP[1]) =	b/0	// b

